tests/k8st/tests/test_bgp_advert.py

```diff
@@ -0,0 +1,217 @@
+                     # flapping since multiple nodes update their BGP
+                     # configuration at the same time, GR is not guaranteed to
+                     # work correctly in this scenario.
+  merge paths on;
+}
+
+# Watch interface up/down events.
+protocol device {
+  debug { states };
+  scan time 2;    # Scan interfaces every 2 seconds
+}
+
+protocol direct {
+  debug { states };
+  interface -"cali*", "*"; # Exclude cali* but include everything else.
+}
+
+# Template for all BGP clients
+template bgp bgp_template {
+  debug { states };
+  description "Connection to BGP peer";
+  local as 64512;
+  multihop;
+  gateway recursive; # This should be the default, but just in case.
+  import all;        # Import all routes, since we don't know what the upstream
+                     # topology is and therefore have to trust the ToR/RR.
+  export all;
+  source address 10.192.0.5;  # The local address we use for the TCP connection
+  add paths on;
+  graceful restart;  # See comment in kernel section about graceful restart.
+  connect delay time 2;
+  connect retry time 5;
+  error wait time 5,30;
+}
+
+# ------------- Node-to-node mesh -------------
+# For peer /host/kube-master/ip_addr_v4
+protocol bgp Mesh_10_192_0_2 from bgp_template {
+  neighbor 10.192.0.2 as 64512;
+  passive on; # Mesh is unidirectional, peer will connect to us.
+}
+
+
+# For peer /host/kube-node-1/ip_addr_v4
+protocol bgp Mesh_10_192_0_3 from bgp_template {
+  neighbor 10.192.0.3 as 64512;
+  passive on; # Mesh is unidirectional, peer will connect to us.
+}
+
+# For peer /host/kube-node-2/ip_addr_v4
+protocol bgp Mesh_10_192_0_4 from bgp_template {
+  neighbor 10.192.0.4 as 64512;
+  passive on; # Mesh is unidirectional, peer will connect to us.
+}
+"""
+
+
+class TestBGPAdvert(TestBase):
+    def setUp(self):
+        # Run tearDown in case anything was left up
+        self.tearDown()
+
+        # # Setup external node: use privileged mode for setting routes
+        subprocess.check_call("docker run -d "
+                              "--privileged "
+                              "--name kube-node-extra "
+                              "--network kubeadm-dind-net "
+                              "mirantis/kubeadm-dind-cluster:v1.10",
+                              shell=True)
+        # Install bird on extra node
+        subprocess.check_call("docker exec kube-node-extra apt update", shell=True)
+        subprocess.check_call("docker exec kube-node-extra apt install -y bird", shell=True)
+        subprocess.check_call("docker exec kube-node-extra mkdir /run/bird", shell=True)
+        with open('bird.conf', 'w') as birdconfig:
+            birdconfig.write(bird_conf)
+        subprocess.check_call("docker cp bird.conf kube-node-extra:/etc/bird/bird.conf", shell=True)
+        subprocess.check_call("rm bird.conf", shell=True)
+        subprocess.check_call("docker exec kube-node-extra service bird restart", shell=True)
```

Comment:
> Also, a key point here is that the extra node must not be a member of the k8s cluster, and in particular not run kube-proxy. Is that the case with this setup?
Yes, it absolutely is.

And in fact I was worried that by using calico-node, we might end up with more integration with the cluster than we were expecting (e.g. bird config).  But happy to do that if you tell me its safe to do.  

The way it is now, we can see *exactly* what's going into that node right here in the test.  Moving that stuff to a Dockerfile moves that config out of this test, which might be OK if we were reusing that container somewhere else, but I don't see that we will (at least for now)

Reply:
OK, I'm persuaded for now.  If we subroutine this and start using it multiple times, it probably will be worth doing the `apt update` and `apt install -y bird` once upfront; but no need to do that now.

Re using calico-node I take your point.  FYI I think there are 3 relevant areas of function, if we used calico-node.
1. Whether the node/host is a k8s member and has kube-proxy.  This would completely undermine the test concept.  It depends entirely on the kubeadm-dind-cluster setup, and you seem confident that the right thing is happening - so good.
2. The BIRD, BIRD config and confd logic inside calico-node.  IMO it would definitely be a _feature_ for us to be using those, not a concern; and I'm confident that they don't in any way undermine what we're trying to test.
3. The Felix inside calico-node.  Theoretically I think this is fine, but I do feel a bit of FUD about whether running Felix might somehow interact with routing and possibly undermine the test intention.

Perhaps we could usefully do some more polishing to facilitate using calico/node in this kind of context; but no need to do that in this particular ST.